keep-network/keep-core
# Support the new rewards mechanism in KEEP token dashboard

Ref: https://github.com/keep-network/keep-ecdsa/pull/630

This PR adds support for the new rewards mechanism described in https://github.com/keep-network/keep-ecdsa/pull/644. In order to have all the required parameters to call the [`claim` function](https://github.com/keep-network/keep-ecdsa/blob/master/solidity/contracts/ECDSARewardsDistributor.sol#L72), the file generated by [merkle distributor](https://github.com/keep-network/keep-ecdsa/pull/636) is necessary. The output file should be in the `src/rewards-allocation/rewards.json` file. The dapp interprets `merkle_root` as a reward interval.

### This PR provides main changes:
* the new tbtc rewards page on which the beneficiary can view rewards history, view available rewards balance, and withdraw rewards (**Note: currently, the data on the tBTC rewards page is displayed from the beneficiary's perspective**),
* integration with `ECDSARewardsDistributor` contract and the output file generated by merkle distributor,
* `Timer` component #2212,
* separate pages for `Random Beacon` and `tBTC` instead of a single `Overview` page,
* update `start_dashboard.sh` script- generating necessary files to test the new rewards mechanism in KEEP token dashbord on the real merkle roots.

### This PR does not provide: 
* the `boost`, `eth score` and `APY` calculation because one beneficiary may have multiple operators and it is unclear how this information should be displayed in such a case,
* `witheld` status - at least for now - this information is not included in the input from reward distribution scripts,
* `accumulating` status - at least for now - computing this value is possible but also time-consuming when rendering the page.

### Release notes
For each interval, the dapp should be released with a file generated by merkle distributor which should contain previous merkle roots (intervals) and the new one. The file should be in the `src/rewards-allocation/rewards.json`. Note the `reward.json` file which is provided by this PR contains mock data and should not be used on the production build.

### Testing locally
To test the new rewards mechanism locally:
* please switch to `deploy-ecdsa-rewards` branch in `keep-ecdsa` repo or make sure the current branch includes files created in https://github.com/keep-network/keep-ecdsa/pull/630,
* run `start_dashbord.sh` script.

<details>
  <summary>Screenshots</summary>
  
![obraz](https://user-images.githubusercontent.com/57687279/102330069-686d6000-3f89-11eb-82fb-3c3010c025a5.png)

![obraz](https://user-images.githubusercontent.com/57687279/102329713-feed5180-3f88-11eb-9f40-cd1f388ca5f6.png)

</details>
